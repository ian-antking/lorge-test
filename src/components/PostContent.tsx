import React, { useState, useEffect } from 'react'
import { useParams } from 'react-router-dom'
import styled from 'styled-components'
import Markdown from './Markdown'

import getPost from '../utils/get-post'
import decodeContent from '../utils/decode-content'

interface Post {
  id: string
  title: string
  author: string
  blurb: string
  published: string
  content: string
}

const ContentContainer = styled.div`
  display: flex;
  justify-content: center;
`

const ContentCard = styled.div`
  width: 80vw;
  padding: 20px;
  border-radius: 5px;
  box-shadow: ${(props) => props.theme.dark} 1.95px 1.95px 2.6px;
  margin: 10px;
`

export default (): React.ReactElement => {
  const [content, setContent] = useState({} as Post)
  const { postId } = useParams()

  useEffect(() => {
    const getContent = async () => {
      const data = await getPost(postId as string)
      setContent(data)
    }

    getContent()
  }, [postId])

  if (typeof content.content === 'string') {
    return (
      <ContentContainer>
        <ContentCard>
          <Markdown content={decodeContent(content.content)} />
        </ContentCard>
      </ContentContainer>
    )
  } else {
    return <h1>Loading...</h1>
  }
}
