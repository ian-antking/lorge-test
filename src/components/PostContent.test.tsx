import PostContent from './PostContent'
import { render, waitForElementToBeRemoved } from '@testing-library/react'
import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { readFile } from 'fs/promises'

import '@testing-library/jest-dom'

interface Post {
  id: string
  title: string
  author: string
  blurb: string
  published: string
  content: string
}

describe('PostContent', () => {
  it('renders the markdown as plain text', async () => {
    const history = createMemoryHistory({
      initialEntries: ['/post/id1']
    })

    const content = `# title

    this is some content
    `

    const buffer: Buffer = Buffer.from(content, 'ascii')

    const data: Post = {
      id: 'id1',
      title: 'title',
      published: new Date(Date.now()).toISOString(),
      blurb: 'blurbText',
      author: 'authorText',
      content: buffer.toString('base64')
    }

    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ post: data })
      })
    ) as jest.Mock

    const { findByText } = render(
      <HistoryRouter history={history}>
        <PostContent />
      </HistoryRouter>
    )

    expect(await findByText('title')).toBeInTheDocument()
    expect(await findByText('this is some content')).toBeInTheDocument()
  })

  it('renders consistently', async () => {
    const history = createMemoryHistory({
      initialEntries: ['/post/id1']
    })

    const content: string = await readFile(
      'src/components/test-data/article.md',
      'utf-8'
    )

    const buffer: Buffer = Buffer.from(content, 'ascii')

    const data: Post = {
      id: 'id1',
      title: 'title',
      published: new Date(Date.now()).toISOString(),
      blurb: 'blurbText',
      author: 'authorText',
      content: buffer.toString('base64')
    }

    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ post: data })
      })
    ) as jest.Mock

    const { asFragment, findByText } = render(
      <HistoryRouter history={history}>
        <PostContent />
      </HistoryRouter>
    )

    waitForElementToBeRemoved(await findByText('Loading...'))

    expect(await findByText('title')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
