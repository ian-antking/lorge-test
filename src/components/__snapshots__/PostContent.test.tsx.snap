// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostContent renders consistently 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ giNOlw"
  >
    <div
      class="sc-gsnTZi bXFjmF"
    >
      <h1>
        title
      </h1>
      

      <p>
        this is some content
      </p>
      

      <h2>
        subtitle
      </h2>
      

      <ul>
        

        <li>
          ul item 1
        </li>
        

        <li>
          ul item 2
        </li>
        

      </ul>
      

      <ol>
        

        <li>
          ol item 1
        </li>
        

        <li>
          ol item 2
        </li>
        

      </ol>
    </div>
  </div>
</DocumentFragment>
`;
